tests/unit/s2n_connection_serialize_test.c

```diff
@@ -473,6 +490,271 @@ int main(int argc, char **argv)
+
+        /* Server resumption configuration */
+        S2N_BLOB_FROM_HEX(ticket_key,
+                "077709362c2e32df0ddc3f0dc47bba63"
+                "90b6c73bb50f9c3122ec844ad7c2b3e5");
+        uint64_t current_time = 0;
+        uint8_t ticket_key_name[16] = "resumption key\0";
+        EXPECT_SUCCESS(resumption_config->wall_clock(resumption_config->sys_clock_ctx, &current_time));
+        EXPECT_SUCCESS(s2n_config_add_ticket_crypto_key(resumption_config, ticket_key_name, strlen((char *) ticket_key_name),
+                ticket_key.data, ticket_key.size, current_time / ONE_SEC_IN_NANOS));
+
+        /* Client is serialized. Can read a session ticket after deserialization. */
+        {
+            DEFER_CLEANUP(struct s2n_connection *client_conn = s2n_connection_new(S2N_CLIENT),
+                    s2n_connection_ptr_free);
+            EXPECT_NOT_NULL(client_conn);
+            DEFER_CLEANUP(struct s2n_connection *server_conn = s2n_connection_new(S2N_SERVER),
+                    s2n_connection_ptr_free);
+            EXPECT_NOT_NULL(server_conn);
+
+            EXPECT_SUCCESS(s2n_connection_set_config(client_conn, resumption_config));
+            EXPECT_SUCCESS(s2n_connection_set_config(server_conn, resumption_config));
+
+            DEFER_CLEANUP(struct s2n_test_io_stuffer_pair io_pair = { 0 }, s2n_io_stuffer_pair_free);
+            EXPECT_OK(s2n_io_stuffer_pair_init(&io_pair));
+            EXPECT_OK(s2n_connections_set_io_stuffer_pair(client_conn, server_conn, &io_pair));
+
+            EXPECT_SUCCESS(s2n_negotiate_test_server_and_client(server_conn, client_conn));
+            EXPECT_EQUAL(s2n_connection_get_actual_protocol_version(server_conn), S2N_TLS13);
+
+            EXPECT_EQUAL(s2n_stuffer_data_available(&cb_session_data), 0);
+
+            /* Client will be serialized before reading the session ticket sent by the server. */
+            uint8_t buffer[S2N_SERIALIZED_CONN_TLS12_SIZE] = { 0 };
+            EXPECT_SUCCESS(s2n_connection_serialize(client_conn, buffer, sizeof(buffer)));
+
+            /* Initialize new client connection and deserialize the connection */
+            DEFER_CLEANUP(struct s2n_connection *new_client_conn = s2n_connection_new(S2N_CLIENT),
+                    s2n_connection_ptr_free);
+            EXPECT_NOT_NULL(new_client_conn);
+            EXPECT_SUCCESS(s2n_connection_deserialize(new_client_conn, buffer, sizeof(buffer)));
+
+            /* Re-set the config on the client connection as it contains the resumption callback */
+            EXPECT_SUCCESS(s2n_connection_set_config(new_client_conn, resumption_config));
+
+            /* Re-initialize IO pipes. We do not wipe the io pair since we need the session ticket
+             * sent by the server from the negotiate call. */
+            EXPECT_OK(s2n_connections_set_io_stuffer_pair(new_client_conn, server_conn, &io_pair));
```

Comment:
This seems dangerous-- it's not an accurate test. Can you just trigger another session ticket?

Reply:
I mean, I do think it's an accurate test. You have to read all TLS messages anyways, if you skip a message you won't be able to decrypt the message. So we can't throw out any messages. I can get the server to send another session ticket, but I'm just pointing out that it isn't necessary since the client has to read the first ticket no matter what.